From 17a6ebf821cf518a743e5c6ba6bcb51a5f58661f Mon Sep 17 00:00:00 2001
From: Huang Chaofan <chaofan.huang@nxp.com>
Date: Tue, 17 Apr 2018 16:28:29 +0800
Subject: [PATCH 3610/5242] MLK-18067[IMX8QXP B0] Deal with the case that
 receive VID_API_EVENT_FINISHED when reset

commit  65a6e587c4adcefb7b161957d73be57bd9fdd6a2 from
https://source.codeaurora.org/external/imx/linux-imx.git

Deal with the case that receive VID_API_EVENT_FINISHED when reset in vpu
B0 decoder driver

Signed-off-by: Huang Chaofan <chaofan.huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mxc/vpu-decoder-b0/vpu_b0.c |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/mxc/vpu-decoder-b0/vpu_b0.c b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
index a2b4f2e..212465f 100755
--- a/drivers/mxc/vpu-decoder-b0/vpu_b0.c
+++ b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
@@ -1728,7 +1728,10 @@ static void vpu_api_event_handler(struct vpu_ctx *ctx, u_int32 uStrIdx, u_int32
 	case VID_API_EVENT_STR_FMT_CHANGE:
 		break;
 	case VID_API_EVENT_FINISHED:
-		v4l2_vpu_send_cmd(ctx, uStrIdx, VID_API_CMD_STOP, 0, NULL);
+		if (ctx->wait_rst_done)
+			vpu_dbg(LVL_ERR, "receive VID_API_EVENT_FINISHED when reset, ignore it\n");
+		else
+			v4l2_vpu_send_cmd(ctx, uStrIdx, VID_API_CMD_STOP, 0, NULL);
 		break;
 	default:
 		break;
@@ -2260,11 +2263,14 @@ static int v4l2_release(struct file *filp)
 		wait_for_completion(&ctx->stop_cmp);
 	}
 
+	dev->ctx[ctx->str_index] = NULL;
 	release_queue_data(ctx);
 	ctrls_delete_decoder(ctx);
 	v4l2_fh_del(&ctx->fh);
 	v4l2_fh_exit(&ctx->fh);
+	mutex_lock(&dev->dev_mutex);
 	clear_bit(ctx->str_index, &dev->instance_mask);
+	mutex_unlock(&dev->dev_mutex);
 
 	for (i = 0; i < MAX_DCP_NUM; i++)
 		if (ctx->dcp_dma_virt[i] != NULL)
-- 
1.7.9.5

